src/index.tsx

```diff
@@ -0,0 +1,269 @@
+  updateItemQuantity: (id: Item["id"], quantity: number) => void;
+  emptyCart: () => void;
+  getItem: (id: Item["id"]) => any | undefined;
+  inCart: (id: Item["id"]) => boolean;
+}
+
+export type Actions =
+  | { type: "SET_ITEMS"; payload: Item[] }
+  | { type: "ADD_ITEM"; payload: Item }
+  | { type: "REMOVE_ITEM"; id: Item["id"] }
+  | {
+      type: "UPDATE_ITEM";
+      id: Item["id"];
+      payload: object;
+    }
+  | { type: "EMPTY_CART" };
+
+export const initialState: any = {
+  items: [],
+  isEmpty: true,
+  totalItems: 0,
+  totalUniqueItems: 0,
+  cartTotal: 0,
+};
+
+const CartContext = React.createContext<CartProviderState | undefined>(
+  initialState
+);
+
+export const createCartIdentifier = (len = 12) =>
+  [...Array(len)].map(() => (~~(Math.random() * 36)).toString(36)).join("");
+
+export const useCart = () => {
+  // This makes sure that the cart functions are always defined before calling it.
+  const context = React.useContext(CartContext);
+  if (!context) throw new Error("Expected to be wrapped in a CartProvider");
+  return context;
+};
+
+function reducer(state: CartProviderState, action: Actions) {
+  switch (action.type) {
+    case "SET_ITEMS":
+      return generateCartState(state, action.payload);
+
+    case "ADD_ITEM": {
+      const items = [...state.items, action.payload];
+
+      return generateCartState(state, items);
+    }
+
+    case "UPDATE_ITEM": {
+      const items = state.items.map((item: Item) => {
+        if (item.id !== action.id) return item;
+
+        return {
+          ...item,
+          ...action.payload,
+        };
+      });
+
+      return generateCartState(state, items);
+    }
+
+    case "REMOVE_ITEM": {
+      const items = state.items.filter((i: Item) => i.id !== action.id);
+
+      return generateCartState(state, items);
+    }
+
+    case "EMPTY_CART":
+      return initialState;
+
+    default:
+      throw new Error("No action specified");
+  }
+}
+
+const generateCartState = (state = initialState, items: Item[]) => {
+  const totalUniqueItems = calculateUniqueItems(items);
+  const isEmpty = totalUniqueItems === 0;
+
+  return {
+    ...initialState,
+    ...state,
+    items: calculateItemTotals(items),
+    totalItems: calculateTotalItems(items),
+    totalUniqueItems,
+    cartTotal: calculateCartTotal(items),
+    isEmpty,
+  };
+};
+
+const calculateItemTotals = (items: Item[]) =>
+  items.map((item) => ({
+    itemTotal: item.price * item.quantity!,
+    ...item,
+  }));
+
+const calculateCartTotal = (items: Item[]) =>
+  items.reduce((total, item) => total + item.quantity! * item.price, 0);
+
+const calculateTotalItems = (items: Item[]) =>
+  items.reduce((sum, item) => sum + item.quantity!, 0);
+
+const calculateUniqueItems = (items: Item[]) => items.length;
+
+export const CartProvider: React.FC<{
+  children?: React.ReactNode;
+  id?: string;
+  defaultItems?: Item[];
+  onSetItems?: any;
+  onItemAdd?: any;
+  onItemUpdate?: any;
+  onItemRemove?: any;
```

Comment:
Is it worth having some type safety around these handlers?

For example, `onItemRemove && onItemRemove(id);` it would be important here rto know the `id` is returned, but in some others it's the items.

Reply:
I added some stronger types, also one for storage.

Reply:
This made my question if the `onItemUpdate && onItemUpdate(payload);` in addItem(), should be changed to `onItemAdd && onItemAdd(payload);`

Reply:
Also wondering if savedCart and saveCart is too similar name wise.